provider/provider.go

```diff
@@ -0,0 +1,136 @@
+package provider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+
+	"github.com/codedellemc/gocsi/csi"
+	"github.com/codedellemc/goioc"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+
+	"github.com/codedellemc/csi-blockdevices/services"
+)
+
+const (
+	nodeEnvVar = "BDPLUGIN_NODEONLY"
+	ctlrEnvVar = "BDPLUGIN_CONTROLLERONLY"
+)
+
+var (
+	errServerStarted = errors.New(
+		services.SpName + ": the server has been started")
+	errServerStopped = errors.New(
+		services.SpName + ": the server has been stopped")
+)
+
+func init() {
+	goioc.Register(services.SpName, newProvider)
+}
+
+type provider struct {
+	sync.Mutex
+	server *grpc.Server
+	closed bool
+	plugin *services.StoragePlugin
+}
+
+func newProvider() interface{} {
+	return &provider{}
+}
+
+// Serve accepts incoming connections on the listener lis, creating
+// a new ServerTransport and service goroutine for each. The service
+// goroutine read gRPC requests and then call the registered handlers
+// to reply to them. Serve returns when lis.Accept fails with fatal
+// errors.  lis will be closed when this method returns.
+// Serve always returns non-nil error.
+func (p *provider) Serve(ctx context.Context, li net.Listener) error {
+	log.WithField("name", services.SpName).Info(".Serve")
+	if err := func() error {
+		p.Lock()
+		defer p.Unlock()
+		if p.closed {
+			return errServerStopped
+		}
+		if p.server != nil {
+			return errServerStarted
+		}
+		p.server = grpc.NewServer()
+		return nil
+	}(); err != nil {
+		return errServerStarted
+	}
+
+	p.plugin = &services.StoragePlugin{}
+	p.plugin.Init()
+
+	// Always host the Identity Service
+	csi.RegisterIdentityServer(p.server, p.plugin)
+
+	_, nodeSvc := os.LookupEnv(nodeEnvVar)
+	_, ctrlSvc := os.LookupEnv(ctlrEnvVar)
+
+	if nodeSvc && ctrlSvc {
+		log.Errorf("Cannot specify both %s and %s",
+			nodeEnvVar, ctlrEnvVar)
+		return fmt.Errorf("Cannot specify both %s and %s",
+			nodeEnvVar, ctlrEnvVar)
+	}
+
+	switch {
+	case nodeSvc:
+		csi.RegisterNodeServer(p.server, p.plugin)
+		log.Debug("Added Node Service")
+	case ctrlSvc:
+		csi.RegisterControllerServer(p.server, p.plugin)
+		log.Debug("Added Controller Service")
+	default:
+		csi.RegisterControllerServer(p.server, p.plugin)
+		log.Debug("Added Controller Service")
+		csi.RegisterNodeServer(p.server, p.plugin)
+		log.Debug("Added Node Service")
+	}
+
+	// start the grpc server
+	if err := p.server.Serve(li); err != grpc.ErrServerStopped {
+		return err
+	}
+	return errServerStopped
```

Comment:
Hi @codenrhoden,

Also, when operating in in-proc mode I wonder if it shouldn't return the standard gRPC stopped error so the host process doesn't have to check for some custom stop error string based on the plug-in or service name?

Reply:
Good point. Will make that change.